docs/NNNN-schema-registry.md

```diff
@@ -0,0 +1,152 @@
+
+We describe a standardised layout for a schema repository - essentially implemented as a 'meta' OCFL object: a sub-directory containing local copies of schemata. This directory contains all schemata, with standard OCFL integrity checks, in use within the OCFL root.
+
+## Parameters
+
+Configuration is done by setting appropriate values in config.json at the top level of the extension's directory. The keys expected are:
+
+extensionName
+: String identifying the extension. Must be "NNNN-schema-registry".
+
+schemaDigestAlgorithm
+: Algorithm used for calculating safe filenames from schema identifiers. "md5" is the default.
+
+Additionally a schema_registry.json file should be present. This contains the keys -
+
+digestAlgorithm
+: Algorithm used for calculating fixity of the manifest and stored schema files.
+
+manifest
+: An OCFL-style manifest of the stored schema files.
+
+schemaMap
+: Provides a convenient mapping of the original schema identifiers to the digest used to identify the schema.
+
+Example files are shown below.
+
+
+### Implementation
+
+In an OCFL root where this extension is configured, tools reading the OCFL objects are not required to be aware of it. However, accessing local copies of schemata is also likely to decrease latency and may provide a small performance benefit. Thus OCFL reading applications relying on schemata access (eg for validation) may wish to consider reading these from the schema registry.
+
+It is important that a system implementing this extension ensures all OCFL objects written are checked for schema references. Where not already registered, the schema is retrieved and registered as described below.
+
+Two possible implementation cases are considered:
+
+* Where an OCFL is being generated as a snapshot for export or backup purposes, it is reasonable to build or check the completeness of the schema registry as each OCFL object is created. Thus the exported root will contain all necessary schemata.
+
+* Where an OCFL provides the persistence layer for a live repository, it is recommended that a background process discovers (or is informed) of new OCFL object/versions. This process can asynchronously process the object contents, retrieve the required schemata and perform necessary registrations with the schema registry.
+
+Since schemata are typically identified by the URL of the specific version referenced, a hashed representation of the identifier is used to create a safe filename for storage. MD5 is likely to be sufficient although other algorithms may be configured. The extension's inentory stores a map of the hashed identifiers, thus allowing easy access to the appropriate schema from unmodified OCFL objects.
+
+### Registering a schema with the extension
+
+* New OCFL objects/versions are inspected for JSON, XML files. 
+* Within these files, references to external schema ($schema / XML-DTD etc) are extracted
+* The local filename is derived using the specified schemaDigestAlgorithm.
```

Comment:
This means, "hash the schema URL using the `schemaDigestAlgorithm`," correct?

Reply:
Yes. I've reworded it like this in the PR below for clarity.